pkgs/servers/sql/postgresql/generic.nix

```
@@ -78,25 +107,37 @@ let
     nativeBuildInputs = [
       makeWrapper
       pkg-config
+      removeReferencesTo
     ]
-      ++ lib.optionals jitSupport [ llvmPackages.llvm.dev nukeReferences patchelf ];
+      ++ lib.optionals jitSupport [ llvmPackages.llvm.dev ];
 
     enableParallelBuilding = true;
 
     separateDebugInfo = true;
 
     buildFlags = [ "world" ];
 
-    # Makes cross-compiling work when xml2-config can't be executed on the host.
-    # Fixed upstream in https://github.com/postgres/postgres/commit/0bc8cebdb889368abdf224aeac8bc197fe4c9ae6
-    env.NIX_CFLAGS_COMPILE = lib.optionalString (olderThan "13") "-I${libxml2.dev}/include/libxml2";
+    # libpgcommon.a and libpgport.a contain all paths returned by pg_config and are linked
+    # into all binaries. However, almost no binaries actually use those paths. The following
+    # flags will remove unused sections from all shared libraries and binaries - including
+    # those paths. This avoids a lot of circular dependency problems with different outputs,
+    # and allows splitting them cleanly.
+    env.CFLAGS = "-fdata-sections -ffunction-sections"
+      + (if stdenv'.cc.isClang then " -flto" else " -fmerge-constants -Wl,--gc-sections")
+      + lib.optionalString stdenv'.isDarwin " -fuse-ld=lld"
+      # Makes cross-compiling work when xml2-config can't be executed on the host.
+      # Fixed upstream in https://github.com/postgres/postgres/commit/0bc8cebdb889368abdf224aeac8bc197fe4c9ae6
+      + lib.optionalString (olderThan "13") " -I${libxml2.dev}/include/libxml2";
+
+    # This could be removed once the upstream issue is resolved:
+    # https://postgr.es/m/flat/427c7c25-e8e1-4fc5-a1fb-01ceff185e5b%40technowledgy.de
+    configureFlagsArray = lib.optionalString (stdenv'.isDarwin && atLeast "16") "LDFLAGS_EX_BE=-Wl,-export_dynamic";
```

Shouldn’t that be an array, not a string? Is there any particular reason why it’s not part of the `configureFlags`?

```suggestion
    configureFlagsArray = lib.optionals (stdenv'.isDarwin && atLeast "16") [ "LDFLAGS_EX_BE=-Wl,-export_dynamic" ];
```

---

This _may_ be fixed with structured attrs being used (not entirely sure), so it may be reasonable to double-check.

---

For structured attributes, configureFlags and configureFlagsArray should be identical.

https://github.com/NixOS/nixpkgs/blob/0360d94ec3386befc146cc7e3986fa17fb9de3b8/pkgs/stdenv/generic/setup.sh#L1295-L1301
https://github.com/NixOS/nixpkgs/blob/0360d94ec3386befc146cc7e3986fa17fb9de3b8/pkgs/stdenv/generic/setup.sh#L286-L290